lib/eqaf.ml

```diff
@@ -212,3 +220,180 @@ let exists_uint8 ?off ~f str =
+  *)
+  let ( - ) , ( + ), ( * ) = Int32.(sub, add, mul) in
+  let ( >> ) = Int32.shift_right_logical in
+  if (m <= 0l) then raise (Invalid_argument "m <= 0") ;
+  if (m >= 16348l) then raise (Invalid_argument "m >= 16348 not supported") ;
+
+  let of_uint32 uint =
+    (* apparently Int64.of_int32 sign-extends ... great... avoid that: *)
+    let b = Bytes.make 8 '\x00' in
+    Bytes.set_int32_le b 0 uint ;
+    Bytes.get_int64_le b 0
+  in
+
+  let x_0 = x in
+
+  let x_2, q_1 =
+    let v = Int32.unsigned_div Int32.min_int m |> of_uint32 in
+    (*let v = 0x80_00_00_00 / m in*) (* floored div *)
+    let x_1, q_0 =
+      let qpart_0 =
+        let open Int64 in
+        shift_right_logical (mul (of_uint32 x_0) v) 31
+        |> to_int32
+      in
+      x_0 - (qpart_0 * m), qpart_0
+    in
+    let qpart_1 =
+      let open Int64 in
+      shift_right_logical (mul (of_uint32 x_1) v) 31
+      |> to_int32 in
+    x_1 - (qpart_1 * m),
+    (q_0 + qpart_1 + 1l) in
+  let x_3 = x_2 - m  in
+  let mask = 0l - (x_3 >> 31) in
+  q_1 + mask, x_3 + (Int32.logand mask m)
+
+let ascii_of_int32 ~digits (n:int32) : string =
+  (* Recursively calls [divmod n 10]; the remainder is turned into ASCII
+     and the quotient is used for the next division.*)
+  if digits < 0 then raise (Invalid_argument "digits < 0");
+  let out = Bytes.make digits '0' in
+  let rec loop x = function
+    | -1 -> Bytes.unsafe_to_string out
+    | idx ->
+      let next, this = divmod ~x ~m:10l in
+      Bytes.set out idx @@ char_chr (0x30 lor (Int32.to_int this)) ;
+      loop next (pred idx)
+  in loop n (pred digits)
+
+let[@inline always] to_hex_nibble f : char =
+  let a = 86 + f in
+  let c = 1 + ((a - 71 * ((a land 0x10) lsr 4)) lor 0x20) in
+  char_chr c
+
+let hex_of_string rawbytes =
+  String.init (2 * String.length rawbytes)
+    (fun idx ->
+       let byt = String.get rawbytes (idx lsr 1) |> Char.code in
+       (* select which 4 bits to use, this can probably be done faster:*)
+       let nib = 0xf land (byt lsr (((lnot idx) land 1) lsl 2)) in
+       to_hex_nibble nib)
+
+let hex_of_bytes rawbytes = hex_of_string (Bytes.unsafe_to_string rawbytes)
+
+let[@inline always] lowercase_hex_nibble_to_int ch : int =
+  (* converts [0-9af] to [int] in range [0-15].
+     NB: THIS FUNCTION DOES NOT HANDLE UPPERCASE [A-F] AND DOES NO
+         ERROR DETECTION. ONLY WORKS FOR [0-9a-f] - but then it's fast :-)
+  *)
+  (* camlExample__nibble_to_int_1199:
+        addq    $-32, %rax
+        andq    $-65, %rax
+        movq    %rax, %rbx
+        andq    $45, %rbx
+        shrq    $4, %rbx
+        orq     $1, %rbx
+        imulq   $71, %rbx
+        subq    %rbx, %rax
+        addq    $71, %rax
+        ret
+  *)
+  let x =
+    let x =
+      let i = Char.code ch in
+      (i + 39 - 1 - 54) in
+    x land (lnot 0x20) in
+  x - (((x land 0x16) lsr 4) * 71)
+let _ =
+  (* ended up not using this because error detection was nicer,
+     if a bit slower. *)
```

Comment:
I guess this should be removed? Is there a case for a fast hex decoder for lowercase only, without error detection?

Reply:
I don't understand why you put this code. It should be removed indeed.

Reply:
I thought it was interesting and didn't want to lose it, but I have moved it to a different repo.